vm/src/pyobject.rs

```diff
@@ -648,6 +648,21 @@ impl PyObject<dyn PyObjectPayload> {
             Err(self)
         }
     }
+
+    /// Downcast this PyObjectRef to an `Rc<PyObject<T>>`. The [`downcast`](#method.downcast) method
+    /// is generally preferred, as the `PyRef<T>` it returns implements `Deref<Target=T>`, and
+    /// therefore can be used similarly to an `&T`.
+    pub fn downcast_generic<T: PyObjectPayload>(
+        self: Rc<Self>,
+    ) -> Result<Rc<PyObject<T>>, PyObjectRef> {
+        if self.payload_is::<T>() {
+            let ptr = Rc::into_raw(self) as *const PyObject<T>;
+            let ret = unsafe { Rc::from_raw(ptr) };
+            Ok(ret)
+        } else {
+            Err(self)
```

Comment:
is there benefits to return `self` for `Err`?

Reply:
It's the same pattern as `PyObject::downcast()` and `Any::downcast()`; it allows the caller to get the object back if the downcasting failed.